src/lmdb.ml

```diff
@@ -572,67 +570,43 @@ module type S = sig
   type elt
 
   val create : ?create:bool -> Env.t -> string -> t
-  val get : t -> key -> elt
-  val put : ?flags:PutFlags.t -> t -> key -> elt -> unit
-  val append : t -> key -> elt -> unit
-  val remove : ?elt:elt -> t -> key -> unit
-
-  module Txn : sig
-
-    type 'a txn constraint 'a = [< `Read | `Write ]
-
-    val go :
-      ?parent:([< `Read | `Write ] as 'a) txn ->
-      rw:'a ->
-      t ->
-      ('a txn -> 'b) -> 'b option
-
-    val abort : 'a txn -> 'b
-    val get : 'a txn -> key -> elt
-    val put : ?flags:PutFlags.t -> [> `Write ] txn -> key -> elt -> unit
-    val append : [> `Write] txn -> key -> elt -> unit
-    val remove : ?elt:elt -> [> `Write ] txn -> key -> unit
-    val env : 'a txn -> Env.t
-    val stats : 'a txn -> Env.stats
-    val compare : 'a txn -> key -> key -> int
-    val compare_key : 'a txn -> key -> key -> int
-    val compare_elt : 'a txn -> elt -> elt -> int
-    val drop : ?delete:bool -> [< `Write ] txn -> unit
-
-  end
-
-  module Cursor : sig
+  val get : ?txn:[> `Read] Txn.t -> t -> key -> elt
+  val put : ?txn:[> `Read] Txn.t -> ?flags:PutFlags.t -> t -> key -> elt -> unit
+  val append : ?txn:[> `Read] Txn.t -> ?flags:PutFlags.t -> t -> key -> elt -> unit
+  val remove : ?txn:[> `Read] Txn.t -> ?elt:elt -> t -> key -> unit
 
+  module Crs : sig
+    type db
     type 'a t constraint 'a = [< `Read | `Write ]
 
-    val go : 'cap Txn.txn -> f:('cap t -> 'a) -> 'a
+    val go : 'c cap -> ?txn:'c Txn.t -> db -> ('c t -> 'a) -> 'a
 
-    val get : _ t -> key * elt
-    val put : ?flags:PutFlags.t -> [> `Write ] t -> key -> elt -> unit
-    val put_here : ?flags:PutFlags.t -> [> `Write ] t -> key -> elt -> unit
-    val remove : ?all:bool -> [> `Write ] t -> unit
+    val get : crs:_ t -> key * elt
+    val put : ?flags:PutFlags.t -> crs:[> `Write ] t -> key -> elt -> unit
+    val put_here : ?flags:PutFlags.t -> crs:[> `Write ] t -> key -> elt -> unit
+    val remove : ?all:bool -> crs:[> `Write ] t -> unit -> unit
 
-    val first : _ t -> key * elt
-    val last : _ t -> key * elt
-    val next : _ t -> key * elt
-    val prev : _ t -> key * elt
+    val first : crs:_ t -> key * elt
+    val last : crs:_ t -> key * elt
+    val next : crs:_ t -> key * elt
+    val prev : crs:_ t -> key * elt
 
-    val seek : _ t -> key -> elt
-    val seek_range : _ t -> key -> elt
+    val seek : crs:_ t -> key -> elt
+    val seek_range : crs:_ t -> key -> elt
 
-    val first_dup : _ t -> key * elt
-    val last_dup : _ t -> key * elt
-    val next_dup : _ t -> key * elt
-    val prev_dup : _ t -> key * elt
+    val first_dup : crs:_ t -> key * elt
+    val last_dup : crs:_ t -> key * elt
+    val next_dup : crs:_ t -> key * elt
+    val prev_dup : crs:_ t -> key * elt
 
-    val seek_dup : _ t -> key -> elt
-    val seek_range_dup : _ t -> key -> elt
+    val seek_dup : crs:_ t -> key -> elt
+    val seek_range_dup : crs:_ t -> key -> elt
 
-  end
+  end with type db := t
```

Comment:
I usually really dislike this idioms, but here, I don't see a nice solution out. I would really prefer not to use it ... :/